File: app/src/test/java/com/mohsenoid/rickandmorty/data/RepositoryQueryEpisodesTest.java
package com.mohsenoid.rickandmorty.data;

import org.json.JSONException;
import org.junit.Test;

import java.io.IOException;

import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

public class RepositoryQueryEpisodesTest extends RepositoryTest {

    @Test
    public void testIfQueryEpisodesCallsApiClientMethodWhenIsOnline() throws IOException, JSONException {
        // GIVEN
        stubConfigProviderIsOnline(true);
        int page = 1;

        // WHEN
        repository.queryEpisodes(page, null);

        // THEN
        verify(db, times(1)).queryAllEpisodes(page);
        verify(networkClient, times(1)).getEpisodes(page);
    }

    @Test
    public void testIfQueryEpisodesCallsApiClientMethodWhenIsOffline() throws IOException, JSONException {
        // GIVEN
        stubConfigProviderIsOnline(false);
        int page = 1;

        // WHEN
        repository.queryEpisodes(page, null);

        // THEN
        verify(db, times(1)).queryAllEpisodes(page);
        verify(networkClient, times(0)).getEpisodes(page);
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/data/RepositoryQueryEpisodesTest.kt
package com.mohsenoid.rickandmorty.data

import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import kotlinx.coroutines.runBlocking
import org.junit.Test

class RepositoryQueryEpisodesTest : RepositoryTest() {

    @Test
    fun `test if queryEpisodes calls networkClient and db when isOnline`() {
        runBlocking {
            // GIVEN
            stubConfigProviderIsOnline(true)
            val page = 1

            // WHEN
            repository.queryEpisodes(page = page, callback = null)

            // THEN
            verify(episodeDao, times(1)).queryAllEpisodes(page)
            verify(networkClient, times(1)).getEpisodes(page)
        }
    }

    @Test
    fun `test if queryEpisodes calls db only when isOffline`() {
        runBlocking {
            // GIVEN
            stubConfigProviderIsOnline(false)
            val page = 1

            // WHEN
            repository.queryEpisodes(page, null)

            // THEN
            verify(episodeDao, times(1)).queryAllEpisodes(page)
            verify(networkClient, times(0)).getEpisodes(page)
        }
    }
}
